install/kubernetes/cilium/charts/nodeinit/templates/daemonset.yaml

```diff
@@ -55,7 +55,7 @@ spec:
                     fi
 
                     echo "Waiting on pods to stop..."
-                    if grep -q 'docker' /etc/crictl.yaml; then
+                    if [ ! -x /etc/crictl.yaml ] || grep -q 'docker' /etc/crictl.yaml; then
```

Comment:
Can you explain why `-x` is used here? 

With your changes, docker will be used for below scenario. However, `crictl` should be executed intead.

```
root@kind-control-plane:/# ls -lrt /etc/crictl.yaml 
-rw-r--r-- 1 root root 56 Jul  8 21:00 /etc/crictl.yaml

root@kind-control-plane:/# cat /etc/crictl.yaml
runtime-endpoint: unix:///run/containerd/containerd.sockroot@kind-control-plane
```

Reply:
> I believe docker is the default if this hasn't been configured at all so if that file doesn't exist then use docker.

I just checked for EKS and GKE clusters that I have, both are having `/etc/crictl.yaml` file and both are using docker.

Might need others to confirm this.

Reply:
According to [this page](https://www.stackrox.com/post/2020/02/eks-vs-gke-vs-aks/), it aligns with what you're saying @sayboras. I'd also like other to weigh in as I'm not too familiar with the node-init area.

Reply:
> Can you explain why `-x` is used here?

I thought this was the "exists" test; I just double checked and it is executable which would be problematic. I'm fixed that to use `-f` which I think is more accurate.

Reply:
What this series of conditions is trying to determine is "What is the container runtime?"  It is currently (and I'm extending this some) using `/etc/crictl.yaml` as a signal, but this is probably not exactly correct.  In a given random installation this config file might even be configured to be elsewhere.

My thinking is to follow the current pattern and try to improve the situation. In most cases if `crictl.yaml` explicitly mentions docker or there is no `crictl.yaml` config at all then the kubelet is _probably_ using docker since that is the current default ([accoding to the docs](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/) on `--container-runtime`).  This change (now fixed with `-f`) will capture how AKS behaves (no crictl file) and I don't think will break EKS/GKE where that file does exist.

Other ideas:
* Move this test down to the bottom of the if condition stack 
* Have the last two conditions test for the presence of the `docker` binary and `crictl` binary and use whichever one is found
* Try to parse kubelet's commandline to look for relevant flags like 
  * `--container-runtime` currently defaults to docker which was my thinking on "if this hasn't been configured, use docker"
 * `--config`, and then parse that
 * I don't really like this option. It will be much harder to implement (error prone) and still doesn't catch that the default value if no flags are specified can change, or be compiled independently to be different.

Reply:
I just confirmed on the 3 AKS scenarios i know about (from [these docs](https://docs.microsoft.com/en-us/azure/aks/cluster-configuration#container-runtime-configuration-preview))
* image `AKS Ubuntu 16.04`:
  * This is the default 
  * `/etc/crictl.yaml` doesn't exist, testing its existence works
  * containers are restarted to get onto CNI net
  * cilium connectivity-check succeeds
* image `AKS Ubuntu 18.04`
  * Built with `--aks-custom-headers "CustomizedUbuntu=aks-ubuntu-1804"`
  * `/etc/crictl.yaml` doesn't exist, testing its existence works
  * containers are restarted to get onto CNI net
  * cilium connectivity-check succeeds
* image `AKS Ubuntu 18.04` w/ containerd
  * built with `--aks-custom-headers "CustomizedUbuntu=aks-ubuntu-1804,ContainerRuntime=containerd"`
  * `/etc/crictl.yaml` is present; it sets `runtime-endpoint: unix:///run/containerd/containerd.sock`
  * `crictl` binary is not present
  * cilium-node-init logs show restart fails

            Restarting kubenet managed pods
            Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
            bash: line 86: crictl: command not found
            bash: line 86: crictl: command not found
            bash: line 86: crictl: command not found
            bash: line 86: crictl: command not found
            bash: line 86: crictl: command not found
            bash: line 86: crictl: command not found
            bash: line 86: crictl: command not found
            bash: line 86: crictl: command not found
            bash: line 86: crictl: command not found
            Node initialization complete
            bash: line 86: crictl: command not found
            !!! startup-script succeeded!

  * Other issues with pods failing to start, haven't fully debugged it yet b/c i think that's outside of the scope of this ticket. W/ crictl not present the restart pods technique node-init is attempting will be harder.

Reply:
Thanks for your detailed explanation and checking. I think your changes now are definitely good enough for the scope of issue. Docker will be used in either of below cases:

- /etc/crictl.yaml doesn't not exists
- Or `docker` is specifically mentioned in /etc/crictl.yaml.

Ortherwise, `critctl` will be used by default.